src/sleep/sleep.c

```diff
@@ -264,77 +264,110 @@ static int execute(
-                        log_debug_errno(r, "Suspend Interval value set to %s: %m", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
-                else
-                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
-
-                r = get_total_suspend_interval(last_capacity, &total_suspend_interval);
-                if (r < 0)
-                        log_debug_errno(r, "Failed to estimate suspend interval using previous discharge rate, ignoring: %m");
-                else
-                        suspend_interval = total_suspend_interval;
-
-                log_debug("Set timerfd wake alarm for %s", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
-                /* Wake alarm for system with or without battery to hibernate or estimate discharge rate whichever is applicable */
-                timespec_store(&ts.it_value, suspend_interval);
+        r = check_wakeup_type(&wakeup_type_bit);
+        if (r < 0)
+                log_debug_errno(r, "Failed to fetch wakeup-type bit: %m");
 
-                if (timerfd_settime(tfd, 0, &ts, NULL) < 0)
-                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+        r = battery_trip_point_alarm_exists();
+        if (r < 0)
+                log_debug_errno(r, "Failed to check acpi_btp support enabled or not: %m");
 
+        if (r > 0 && wakeup_type_bit > 0) {
+                log_debug("Attempting to suspend...");
                 r = execute(sleep_config, SLEEP_SUSPEND, NULL);
                 if (r < 0)
                         return r;
 
-                r = fd_wait_for_event(tfd, POLLIN, 0);
+                r = check_wakeup_type(&wakeup_type_bit);
                 if (r < 0)
-                        return log_error_errno(r, "Error polling timerfd: %m");
-                /* Store fd_wait status */
-                woken_by_timer = FLAGS_SET(r, POLLIN);
+                        return log_debug_errno(r, "Failed to fetch wakeup-type bit: %m");
 
-                r = fetch_batteries_capacity_by_name(&current_capacity);
-                if (r < 0) {
-                        /* In case of no battery or error while getting charge level, no need to measure
-                         * discharge rate. Instead system should wakeup if it is manual wakeup or
-                         * hibernate if this is a timer wakeup.   */
-                        log_debug_errno(r, "Battery capacity percentage unavailable, cannot estimate discharge rate: %m");
-                        if (!woken_by_timer)
-                                return 0;
-                        break;
-                }
+                if (wakeup_type_bit != 3)
+                        /* For APM Timer wakeup, system should hibernate else wakeup */
+                        return 0;
+        }
+        else {
+                while (battery_is_low() == 0) {
```

Comment:
Is this `else` really necessary?? What happens `wakeup_type_bit == 3` in the above?

If the logic is correct, then please move `else` after `}`, and drop unnecessary `{`:
```
} else
        while (battery_is_low() == 0)
```

Reply:
same as above.
ok. will fix else